index.bs

```diff
@@ -1,164 +1,194 @@
 
+ed_field       ::= ed_field_kind ed_field_value
 
-# Relationship with WebAssembly code # {#webassembly-code}
+ed_field_kind  ::= 0 ;; URL
+                 | 1 ;; File path
 
-The DWARF information can refer to an instruction pointer or location in memory. In addition to that, special WebAssembly primitives such as locals or globals need to be referenced.
+ed_field_value ::= b∗:<a href="https://webassembly.github.io/spec/core/binary/conventions.html#binary-vec">vec</a>(<a href="https://webassembly.github.io/spec/core/binary/values.html#binary-byte">byte</a>)                      (if <a href="https://webassembly.github.io/spec/core/binary/values.html#binary-utf8">utf8</a>(ed_field_value) = b*)
+</pre>
+
+If `ed_field_kind` is `0`, then the `ed_field_value` UTF-8 string must be a URL pointing to the external DWARF file. If `ed_field_kind` is `1`, then the `ed_field_value` UTF-8 string must be a file path pointing to the external DWARF file. If there are multiple `ed_field`s, a DWARF consumer may choose any of them.
+
+Note: Being explicit about whether the external DWARF file is referenced by URL or by file path simplifies the job of getting the external DWARF file for DWARF consumers. Something that has plagued the source map tooling community has been inconsistent use of URLs vs file paths in different parts of the pipeline and tools that always assume one or the other.
+
+Note: Allowing multiple `ed_field`s allows tools to specify both a file path for local development and URL for development on a server. A command line tool might prefer to use file paths, while browser devtools might prefer to use a URL, even if these ultimately point to the same external DWARF file.
 
-## Instruction Pointer ## {#pc}
+The external DWARF file pointed to by an `ed_field` must be a WebAssembly file with [[#embedding-DWARF|embedded DWARF debug info]]. A DWARF consumer should ignore any non-DWARF WebAssembly sections in this file.
 
-The instruction pointer (or PC) is currently defined as WebAssembly bytecode offset, starting from code section body.  The mapping instruction pointer to bytecode offset simplifies resolution between the call stack PC pointer and WebAssembly code debug information [[wasm-debugging-capabilities/#code-to-source|ᴿcode-to-source]]. Many sections, such as `.debug_info`, `.debug_line`, etc., refer to a particular instruction by their bytecode offset, and it will be unique.
+*TODO: should we allow pointing to an external DWARF file that does not have embedded DWARF, but instead points to another external DWARF file? If so, we should mention something about allowing consumer implementations to enforce their own arbitrary max depth, and also say something about cycles.*
 
-It is also trivial to map WebAssembly bytecode offset to the derived native PC, e.g. during AOT or JIT compilation, for a use in a native debugger. 
+*TODO: should we allow heterogeneous embedded and external DWARF? Eg, have debug_line embedded, but debug_info external?*
```

Comment:
@dlsniper Is this something you'd be interested in chiming in about? :smile:

Reply:
AFAIK gcc will always include .eh_frame with the main executable because it's used by the exception mechanism. That's already a case where part of the debug info is in the executable and part is outside.

Reply:
>  gcc will always include .eh_frame with the main executable ... [t]hat's already a case where part of the debug info is in the executable

@aarzilli I don't see .eh_frame being part of DWARF specification, unless I'm missing something, there is somewhat similar .debug_frame.

There is https://github.com/WebAssembly/exception-handling proposal of exception handling during execution in the WebAssembly, and does not look it will rely on .eh_frame-like data, but using try/catch operators.

Reply:
It makes sense the eh_frame is always embedded, since it is required for unwinding the call stack for native executables.

On the other hand, I think eh_frame/debug_frame are irrelevant for wasm. The engine is responsible for maintaining the call stack, not the wasm code, and therefore a debugger doesn't need the code's debug info to get the call stack, it needs debugging APIs from the engine.

It is probably worth adding a section to this document about how eh_frame and debug_frame should be omitted for wasm.